crates/subspace-farmer/src/single_plot_farm.rs

```diff
@@ -687,18 +691,43 @@ impl SinglePlotFarm {
             total_pieces,
         };
 
-        let single_plot_plotter = self.plotter();
-        let span = self.span.clone();
+        let plotter = VerifyingPlotter {
+            single_plot_plotter: self.plotter(),
+            span: self.span.clone(),
+            verification_client,
+        };
+
+        dsn::sync(self.node.clone(), options, plotter)
+    }
+}
+
+// Defines actions on receiving pieces batch. It verifies the pieces against the blockchain and
+// plots them.
+struct VerifyingPlotter<RC> {
+    // RPC client for verification
+    verification_client: Option<RC>,
+    span: Span,
+    single_plot_plotter: SinglePlotPlotter,
+}
+
+#[async_trait]
+impl<RC: RpcClient> OnSync for VerifyingPlotter<RC> {
+    async fn on_pieces(
+        &self,
+        pieces: FlatPieces,
+        piece_indices: Vec<PieceIndex>,
+    ) -> anyhow::Result<()> {
+        let _guard = self.span.enter();
 
-        dsn::sync(self.node.clone(), options, move |pieces, piece_indexes| {
-            let _guard = span.enter();
+        if let Some(ref verification_client) = self.verification_client {
+            verify_pieces_at_blockchain(verification_client, &piece_indices, &pieces).await?;
+        }
 
-            single_plot_plotter
-                .plot_pieces(PiecesToPlot {
-                    pieces,
-                    piece_indexes,
-                })
-                .map_err(Into::into)
-        })
+        self.single_plot_plotter
+            .plot_pieces(PiecesToPlot {
+                pieces,
+                piece_indexes: piece_indices,
```

Comment:
Either use `piece_indexes` for both or rename to `piece_indices`, this looks weird :slightly_smiling_face: